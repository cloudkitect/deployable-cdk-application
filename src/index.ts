import { Task } from 'projen';
import { AwsCdkTypeScriptApp, AwsCdkTypeScriptAppOptions } from 'projen/lib/awscdk';
import { Job, JobPermission, JobStep } from 'projen/lib/github/workflows-model';
import { NodePackageManager, NodeProject } from 'projen/lib/javascript';

export type DeploymentMethod = 'direct' | 'change-set' | 'prepare-change-set';
export type AccountType = 'Dev' | 'Test' | 'QA' | 'Uat' | 'PreProd' | 'Prod' ;

export interface ReleaseConfig {
  /**
   * Environment name to deploy to
   * @example 'Dev' | 'Test' | 'QA' | 'Uat' | 'PreProd' | 'Prod'
   */
  readonly accountType: AccountType | string;
  /**
   * ARN of AWS role to be assumed
   * @example arn:aws:iam::ACCOUNTID:role/ROLENAME
   */
  readonly roleToAssume: string;
  /**
   * Default AWS region for the account
   * @example "us-east-1"
   */
  readonly region: string;
  /**
   * Deploy method
   * @example 'direct' | 'change-set' | 'prepare-change-set'
   * @default 'change-set'
   */
  readonly deploymentMethod?: DeploymentMethod;
  /**
   * Duration of assume role session
   * @default 900
   */
  readonly assumeRoleDurationInSeconds?: number;
  /**
   * Hotswap deployment
   * @default false
   */
  readonly hotswap?: boolean;
  /**
   * Is manual approval required for deployments
   * @default false
   */
  readonly manualApprovalRequired?: boolean;
  /**
   * Comma separated list of github usernames who need to approve the deployments
   */
  readonly approvers?: string;
  /**
   * Deployment tag in the form of v1.0.0
   * @default latest tag
   */
  readonly deploymentTag?: string;
  /**
   * Pre deployment job steps
   */
  readonly preDeploymentSteps?: JobStep[];
  /**
   * Post deployment job steps
   */
  readonly postDeploymentSteps?: JobStep[];

  /**
   * Workflow name where the deployment job should be added.
   * Must be either release or build
   * @default release
   */
  readonly workflowName?: string;
}

/**
 * Deployable cdk application options
 */
export interface DeployableCdkApplicationOptions extends AwsCdkTypeScriptAppOptions {
  /**
   * Stack Pattern representing stacks to be deployed
   */
  readonly stackPattern?: string;
  /**
   * List of release configurations, this will specify environment specific release configurations.
   */
  readonly releaseConfigs: ReleaseConfig[];

}

/**
 * Deployable cdk application
 * Uses PNPM package manager by default
 * It also created deployment tasks for each environment
 */
export class DeployableCdkApplication extends AwsCdkTypeScriptApp {

  readonly deploymentTasks: Task[];
  readonly releaseConfigs: ReleaseConfig[];

  constructor(options: DeployableCdkApplicationOptions) {
    super({
      ...options,
      packageManager: NodePackageManager.PNPM,
      pnpmVersion: '9',
      release: options.release ?? true,
      pullRequestTemplateContents: [
        '## What is this PR for?',
        '',
        '## What type of PR is it?',
        '',
        '- [ ] Bug fix',
        '- [ ] Feature',
        '- [ ] Documentation update',
        '- [ ] Other, please describe:',
        '',
        '## What is the new behavior?',
        '',
        '## Does this PR introduce a breaking change?',
        '',
        '- [ ] Yes',
        '- [ ] No',
        '',
        '## Other information',
        '',
        '## Checklist:',
        '',
        '- [ ] Code review',
        '- [ ] Tests',
        '- [ ] Documentation',
        '',
      ],
    });
    this.releaseConfigs = options.releaseConfigs ?? [];
    this.deploymentTasks=[];
    this.addDevDeps('@cloudkitect/deployable-cdk-app');
    this.createDeploymentTasks(options);
  }

  synth() {
    this.buildDeploymentJobs();
    super.synth();
  }

  createDeploymentTasks(options: DeployableCdkApplicationOptions) {

    for (let releaseConfig of this.releaseConfigs) {
      const deployCommand = this.buildDeployCommand(releaseConfig, options.stackPattern);
      const taskName = `deploy:${releaseConfig.accountType}`;
      const task = this.addTask(taskName, {
        exec: deployCommand,
      });
      this.deploymentTasks.push(task);
    }
  }

  buildDeployCommand(releaseConfig: ReleaseConfig, stackPattern?: string): string {
    const deployMethod = releaseConfig.deploymentMethod ?? 'change-set';
    let command = 'cdk deploy';
    command += stackPattern ? ' ' + stackPattern : ' --all';
    command += releaseConfig.hotswap ? ' --hotswap' : '';
    command += ` --context env=${releaseConfig.accountType}`;
    command += ` --method ${deployMethod}`;
    command += ' --require-approval never';
    return command;
  }

  buildDeploymentJobs() {
    let buildDeploymentNeeds = ['build'];
    let releaseDeploymentNeeds = ['release_github'];
    this.releaseConfigs.forEach((releaseConfig) => {
      const jobDefinition: Job = {
        runsOn: ['ubuntu-latest'],
        needs: releaseConfig.workflowName == 'build' ? buildDeploymentNeeds : releaseDeploymentNeeds,
        permissions: {
          contents: JobPermission.WRITE,
          deployments: JobPermission.READ,
          idToken: JobPermission.WRITE,
          issues: JobPermission.WRITE,
        },
        steps: [],
      };

      jobDefinition.steps.push(this.currentBranch());
      jobDefinition.steps.push(this.checkoutStep('${{ env.CURRENT_BRANCH }}'));
      if (releaseConfig.workflowName != 'build') {
        jobDefinition.steps.push(this.latestTag(releaseConfig));
        jobDefinition.steps.push(this.checkoutStep('${{ env.CURRENT_TAG }}'));
      }
      if (releaseConfig.manualApprovalRequired) {
        jobDefinition.steps.push(this.generateToken());
        jobDefinition.steps.push(this.manualApprovalStep(releaseConfig));
      }
      jobDefinition.steps.push(...(this.package.project as NodeProject).renderWorkflowSetup());
      jobDefinition.steps.push(this.awsCredentials(releaseConfig));
      const preDeploymentSteps = releaseConfig.preDeploymentSteps ?? [];
      for (const steps of preDeploymentSteps) {
        jobDefinition.steps.push(steps);
      }
      jobDefinition.steps.push(this.deploymentStep(this.package.packageManager, releaseConfig));
      const postDeploymentSteps = releaseConfig.postDeploymentSteps ?? [];
      for (const steps of postDeploymentSteps) {
        jobDefinition.steps.push(steps);
      }
      let jobName = `deploy_to_${releaseConfig.accountType}`;
      const job: Record<string, Job> = {};
      job[jobName] = jobDefinition;
      if (releaseConfig.workflowName == 'build') {
        this.buildWorkflow?.addPostBuildJob(jobName, jobDefinition);
        buildDeploymentNeeds = [jobName];
      } else {
        this.release?.addJobs(job);
        releaseDeploymentNeeds = [jobName];
      }
    });
  }

  checkoutStep(passedRef: string): JobStep {
    return {
      name: 'Checkout',
      uses: 'actions/checkout@v4',
      with: {
        'ref': passedRef,
        // 'token': '${{ secrets.PAT }}',
        'fetch-depth': 0,
      },
    };
  }

  awsCredentials(releaseOption: ReleaseConfig): JobStep {
    return {
      name: 'Assume AWS Role',
      uses: 'aws-actions/configure-aws-credentials@v4',
      with: {
        'role-to-assume': releaseOption.roleToAssume,
        'aws-region': releaseOption.region,
        'role-duration-seconds': releaseOption.assumeRoleDurationInSeconds,
        'role-session-name': `${releaseOption.accountType}Session`,
      },
    };
  }

  generateToken(): JobStep {
    return {
      name: 'Generate token',
      id: 'generate_token',
      uses: 'tibdex/github-app-token@v1',
      with: {
        app_id: '${{ secrets.APP_ID }}',
        private_key: '${{ secrets.APP_PRIVATE_KEY }}',
      },
    };
  }

  latestTag(releaseConfig: ReleaseConfig): JobStep {
    let tagCommand = releaseConfig.deploymentTag ?? '$(git describe --tags --abbrev=0 2>/dev/null)';
    console.log(tagCommand);
    return {
      name: 'Get latest tag',
      id: 'get_tag',
      run: 'CURRENT_TAG=$(git describe --tags --abbrev=0 2>/dev/null)\necho "CURRENT_TAG=$CURRENT_TAG" >> $GITHUB_ENV',
    };
  }

  currentBranch(): JobStep {
    return {
      name: 'Get current branch',
      id: 'get_branch',
      run: 'CURRENT_BRANCH=$(git branch --show-current)\necho "CURRENT_BRANCH=$CURRENT_BRANCH" >> $GITHUB_ENV',
    };
  }

  manualApprovalStep(releaseConfig: ReleaseConfig): JobStep {
    return {
      name: 'Wait for approval',
      uses: 'trstringer/manual-approval@v1',
      with: {
        'secret': '${{ steps.generate_token.outputs.token }}',
        'approvers': releaseConfig.approvers,
        'minimum-approvals': 1,
        'issue-title': 'Deployment approval for ${{ env.CURRENT_TAG }} to ' + releaseConfig.accountType,
        'issue-body': 'Please approve or deny the deployment of version ${{ env.CURRENT_TAG }} to' + + releaseConfig.accountType,
      },
    };
  }

  packageManagerCommand(packageManager: NodePackageManager): string {
    if (packageManager === NodePackageManager.NPM) {return 'npm run';}

    if (packageManager === NodePackageManager.YARN_CLASSIC) {return 'yarn';}

    if (packageManager === NodePackageManager.PNPM) {return 'pnpm';}

    throw new Error(`Invalid package manager selected (${packageManager})`);
  }

  deploymentStep(packageManager: NodePackageManager, releaseConfig: ReleaseConfig): JobStep {
    return {
      name: `Deployment to ${releaseConfig.accountType}`,
      run: `${this.packageManagerCommand(packageManager)} deploy:${releaseConfig.accountType}`,
    };
  }

}