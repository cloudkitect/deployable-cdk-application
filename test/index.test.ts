import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import { synthSnapshot } from 'projen/lib/util/synth';
import { DeployableCdkApplication } from '../src';


function mkdtemp() {
  return fs.mkdtempSync(
    path.join(os.tmpdir(), 'projen-test-'),
  );
}

const releaseWorkflowFilePath = '.github/workflows/release.yml';
const deployToProdWorkflowFilePath = '.github/workflows/prod-deployment-workflow.yml';
const deployToUatWorkflowFilePath = '.github/workflows/uat-deployment-workflow.yml';
const buildWorkflowFilePath = '.github/workflows/build.yml';

describe('No stack pattern', () => {
  const project = new DeployableCdkApplication({
    name: 'my-test-app',
    defaultReleaseBranch: 'main',
    cdkVersion: '1.49.0',
    workflowNodeVersion: '14.18.1',
    projenrcTs: true,
    outdir: mkdtemp(),
    releaseConfigs: [{
      accountType: 'Dev',
      deploymentMethod: 'change-set',
      roleToAssume: 'role',
      region: 'us-east-1',
      workflowType: 'build',
    },
    {
      accountType: 'Test',
      deploymentMethod: 'change-set',
      roleToAssume: 'role',
      region: 'us-east-1',
      workflowType: 'release',
    },
    {
      accountType: 'Uat',
      deploymentMethod: 'prepare-change-set',
      roleToAssume: 'role2',
      region: 'us-east-1',
      workflowType: 'manual',
    },
    {
      accountType: 'Prod',
      deploymentMethod: 'prepare-change-set',
      roleToAssume: 'role2',
      region: 'us-east-1',
      workflowType: 'manual',
    }],
    codeArtifactConfig: {
      roleToAssume: 'role',
      region: 'us-east-1',
      accountId: '123',
      repository: 'repo',
      domain: 'domain',
    },
  });
  const synthOutput = synthSnapshot(project);
  test('release workflow', () => {
    expect(synthOutput[releaseWorkflowFilePath]).toMatchSnapshot();
  });
  test('deploy to uat workflow', () => {
    expect(synthOutput[deployToUatWorkflowFilePath]).toMatchSnapshot();
  });
  test('deploy to prod workflow', () => {
    expect(synthOutput[deployToProdWorkflowFilePath]).toMatchSnapshot();
  });
  test('build workflow', () => {
    expect(synthOutput[buildWorkflowFilePath]).toMatchSnapshot();
  });
});
