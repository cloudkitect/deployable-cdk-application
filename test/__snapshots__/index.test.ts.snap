// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`No stack pattern build workflow 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".

name: build
on:
  pull_request: {}
  workflow_dispatch: {}
jobs:
  build:
    runs-on: ubuntu-latest
    permissions:
      contents: write
      id-token: write
    outputs:
      self_mutation_happened: \${{ steps.self_mutation.outputs.self_mutation_happened }}
    env:
      CI: "true"
    steps:
      - name: Checkout
        uses: actions/checkout@v4
        with:
          ref: \${{ github.event.pull_request.head.ref }}
          repository: \${{ github.event.pull_request.head.repo.full_name }}
      - name: Setup pnpm
        uses: pnpm/action-setup@v3
        with:
          version: "9"
      - name: Setup Node.js
        uses: actions/setup-node@v4
        with:
          node-version: 14.18.1
      - name: Install dependencies
        run: pnpm i --no-frozen-lockfile
      - name: build
        run: npx projen build
      - name: Assume AWS Role
        uses: aws-actions/configure-aws-credentials@v4
        with:
          role-to-assume: role
          aws-region: us-east-1
          role-session-name: DevSession
      - name: Deployment to Dev
        run: pnpm deploy:Dev
      - name: Find mutations
        id: self_mutation
        run: |-
          git add .
          git diff --staged --patch --exit-code > .repo.patch || echo "self_mutation_happened=true" >> $GITHUB_OUTPUT
        working-directory: ./
      - name: Upload patch
        if: steps.self_mutation.outputs.self_mutation_happened
        uses: actions/upload-artifact@v4
        with:
          name: .repo.patch
          path: .repo.patch
          overwrite: true
      - name: Fail build on mutation
        if: steps.self_mutation.outputs.self_mutation_happened
        run: |-
          echo "::error::Files were changed during build (see build log). If this was triggered from a fork, you will need to update your branch."
          cat .repo.patch
          exit 1
  self-mutation:
    needs: build
    runs-on: ubuntu-latest
    permissions:
      contents: write
    if: always() && needs.build.outputs.self_mutation_happened && !(github.event.pull_request.head.repo.full_name != github.repository)
    steps:
      - name: Checkout
        uses: actions/checkout@v4
        with:
          token: \${{ secrets.PROJEN_GITHUB_TOKEN }}
          ref: \${{ github.event.pull_request.head.ref }}
          repository: \${{ github.event.pull_request.head.repo.full_name }}
      - name: Download patch
        uses: actions/download-artifact@v4
        with:
          name: .repo.patch
          path: \${{ runner.temp }}
      - name: Apply patch
        run: '[ -s \${{ runner.temp }}/.repo.patch ] && git apply \${{ runner.temp }}/.repo.patch || echo "Empty patch. Skipping."'
      - name: Set git identity
        run: |-
          git config user.name "github-actions"
          git config user.email "github-actions@github.com"
      - name: Push changes
        env:
          PULL_REQUEST_REF: \${{ github.event.pull_request.head.ref }}
        run: |-
          git add .
          git commit -s -m "chore: self mutation"
          git push origin HEAD:$PULL_REQUEST_REF
"
`;

exports[`No stack pattern deploy to prod workflow 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".

name: prod-deployment-workflow
on:
  workflow_dispatch:
    inputs:
      tag:
        description: Version tag to deploy to Prod
        required: true
jobs:
  deploy_to_Prod:
    runs-on: ubuntu-latest
    permissions:
      contents: write
      deployments: read
      id-token: write
      issues: write
    steps:
      - name: Checkout
        uses: actions/checkout@v4
        with:
          ref: main
          fetch-depth: 0
      - name: Get tag
        id: get_tag
        run: |-
          if [ -z "\${{ github.event.inputs.tag }}" ]; then 
           CURRENT_TAG=$(git describe --tags $(git rev-list --tags --max-count=1)) 
           echo "CURRENT_TAG=$CURRENT_TAG" >> $GITHUB_ENV 
          else 
           echo "CURRENT_TAG=\${{ github.event.inputs.tag }}" >> $GITHUB_ENV 
          fi
      - name: Checkout
        uses: actions/checkout@v4
        with:
          ref: \${{ env.CURRENT_TAG }}
          fetch-depth: 0
      - name: Setup pnpm
        uses: pnpm/action-setup@v3
        with:
          version: "9"
      - name: Setup Node.js
        uses: actions/setup-node@v4
        with:
          node-version: 14.18.1
      - name: Install dependencies
        run: pnpm i --frozen-lockfile
      - name: Assume AWS Role
        uses: aws-actions/configure-aws-credentials@v4
        with:
          role-to-assume: role2
          aws-region: us-east-1
          role-session-name: ProdSession
      - name: Deployment to Prod
        run: pnpm deploy:Prod
"
`;

exports[`No stack pattern deploy to uat workflow 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".

name: uat-deployment-workflow
on:
  workflow_dispatch:
    inputs:
      tag:
        description: Version tag to deploy to Uat
        required: true
jobs:
  deploy_to_Uat:
    runs-on: ubuntu-latest
    permissions:
      contents: write
      deployments: read
      id-token: write
      issues: write
    steps:
      - name: Checkout
        uses: actions/checkout@v4
        with:
          ref: main
          fetch-depth: 0
      - name: Get tag
        id: get_tag
        run: |-
          if [ -z "\${{ github.event.inputs.tag }}" ]; then 
           CURRENT_TAG=$(git describe --tags $(git rev-list --tags --max-count=1)) 
           echo "CURRENT_TAG=$CURRENT_TAG" >> $GITHUB_ENV 
          else 
           echo "CURRENT_TAG=\${{ github.event.inputs.tag }}" >> $GITHUB_ENV 
          fi
      - name: Checkout
        uses: actions/checkout@v4
        with:
          ref: \${{ env.CURRENT_TAG }}
          fetch-depth: 0
      - name: Setup pnpm
        uses: pnpm/action-setup@v3
        with:
          version: "9"
      - name: Setup Node.js
        uses: actions/setup-node@v4
        with:
          node-version: 14.18.1
      - name: Install dependencies
        run: pnpm i --frozen-lockfile
      - name: Assume AWS Role
        uses: aws-actions/configure-aws-credentials@v4
        with:
          role-to-assume: role2
          aws-region: us-east-1
          role-session-name: UatSession
      - name: Deployment to Uat
        run: pnpm deploy:Uat
"
`;

exports[`No stack pattern release workflow 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".

name: release
on:
  push:
    branches:
      - main
  workflow_dispatch: {}
jobs:
  release:
    runs-on: ubuntu-latest
    permissions:
      contents: write
    outputs:
      latest_commit: \${{ steps.git_remote.outputs.latest_commit }}
      tag_exists: \${{ steps.check_tag_exists.outputs.exists }}
    env:
      CI: "true"
    steps:
      - name: Checkout
        uses: actions/checkout@v4
        with:
          fetch-depth: 0
      - name: Set git identity
        run: |-
          git config user.name "github-actions"
          git config user.email "github-actions@github.com"
      - name: Setup pnpm
        uses: pnpm/action-setup@v3
        with:
          version: "9"
      - name: Setup Node.js
        uses: actions/setup-node@v4
        with:
          node-version: 14.18.1
      - name: Install dependencies
        run: pnpm i --frozen-lockfile
      - name: release
        run: npx projen release
      - name: Check if version has already been tagged
        id: check_tag_exists
        run: |-
          TAG=$(cat dist/releasetag.txt)
          ([ ! -z "$TAG" ] && git ls-remote -q --exit-code --tags origin $TAG && (echo "exists=true" >> $GITHUB_OUTPUT)) || (echo "exists=false" >> $GITHUB_OUTPUT)
          cat $GITHUB_OUTPUT
      - name: Check for new commits
        id: git_remote
        run: |-
          echo "latest_commit=$(git ls-remote origin -h \${{ github.ref }} | cut -f1)" >> $GITHUB_OUTPUT
          cat $GITHUB_OUTPUT
      - name: Backup artifact permissions
        if: \${{ steps.git_remote.outputs.latest_commit == github.sha }}
        run: cd dist && getfacl -R . > permissions-backup.acl
        continue-on-error: true
      - name: Upload artifact
        if: \${{ steps.git_remote.outputs.latest_commit == github.sha }}
        uses: actions/upload-artifact@v4
        with:
          name: build-artifact
          path: dist
          overwrite: true
  release_github:
    name: Publish to GitHub Releases
    needs: release
    runs-on: ubuntu-latest
    permissions:
      contents: write
    if: needs.release.outputs.tag_exists != 'true' && needs.release.outputs.latest_commit == github.sha
    steps:
      - uses: actions/setup-node@v4
        with:
          node-version: 14.18.1
      - name: Download build artifacts
        uses: actions/download-artifact@v4
        with:
          name: build-artifact
          path: dist
      - name: Restore build artifact permissions
        run: cd dist && setfacl --restore=permissions-backup.acl
        continue-on-error: true
      - name: Release
        env:
          GITHUB_TOKEN: \${{ secrets.GITHUB_TOKEN }}
          GITHUB_REPOSITORY: \${{ github.repository }}
          GITHUB_REF: \${{ github.sha }}
        run: errout=$(mktemp); gh release create $(cat dist/releasetag.txt) -R $GITHUB_REPOSITORY -F dist/changelog.md -t $(cat dist/releasetag.txt) --target $GITHUB_REF 2> $errout && true; exitcode=$?; if [ $exitcode -ne 0 ] && ! grep -q "Release.tag_name already exists" $errout; then cat $errout; exit $exitcode; fi
  deploy_to_Test:
    needs: release_github
    runs-on: ubuntu-latest
    permissions:
      contents: write
      deployments: read
      id-token: write
      issues: write
    steps:
      - name: Checkout
        uses: actions/checkout@v4
        with:
          ref: main
          fetch-depth: 0
      - name: Get tag
        id: get_tag
        run: |-
          if [ -z "\${{ github.event.inputs.tag }}" ]; then 
           CURRENT_TAG=$(git describe --tags $(git rev-list --tags --max-count=1)) 
           echo "CURRENT_TAG=$CURRENT_TAG" >> $GITHUB_ENV 
          else 
           echo "CURRENT_TAG=\${{ github.event.inputs.tag }}" >> $GITHUB_ENV 
          fi
      - name: Checkout
        uses: actions/checkout@v4
        with:
          ref: \${{ env.CURRENT_TAG }}
          fetch-depth: 0
      - name: Setup pnpm
        uses: pnpm/action-setup@v3
        with:
          version: "9"
      - name: Setup Node.js
        uses: actions/setup-node@v4
        with:
          node-version: 14.18.1
      - name: Install dependencies
        run: pnpm i --frozen-lockfile
      - name: Assume AWS Role
        uses: aws-actions/configure-aws-credentials@v4
        with:
          role-to-assume: role
          aws-region: us-east-1
          role-session-name: TestSession
      - name: Deployment to Test
        run: pnpm deploy:Test
"
`;
